### ISSUE TITLE
Unexpected alt/control/shift/system key state in KeyEvent

### ISSUE THREAD

**Issue by pinam45** [SFML/SFML#1506]
## Unexpected alt/control/shift/system key state in KeyEvent

When i press the shift key i get an sf::Event with (event.type == sf::Event::KeyPressed) and (event.key.shift == false) and when i release the key i get an sf::Event with (event.type == sf::Event::KeyReleased) and (event.key.shift == true).

As the doc is:
```cpp
    struct KeyEvent
    {
        Keyboard::Key code;    ///< Code of the key that has been pressed
        bool          alt;     ///< Is the Alt key pressed?
        bool          control; ///< Is the Control key pressed?
        bool          shift;   ///< Is the Shift key pressed?
        bool          system;  ///< Is the System key pressed?
    };
```

Shouldn't i get (event.type == sf::Event::KeyPressed) and (event.key.shift == **true**) when i press the key and (event.type == sf::Event::KeyReleased) and (event.key.shift == **false**) when i release the key (as it is no longer pressed)?

## Environment

OS: Debian GNU/Linux buster/sid:
```
$ uname -a
Linux Maxime-debian 4.18.0-2-amd64 #1 SMP Debian 4.18.10-2 (2018-10-07) x86_64 GNU/Linux
```
Desktop manager: XFCE4
```
$ xfce4-about --version
xfce4-about 4.12.1 (Xfce 4.12)
```
Window manager: i3
```
$ i3 --version
i3 version 4.15 (2018-03-10) © 2009 Michael Stapelberg and contributors
```
SFML:
```
2.5.1 (2f11710abc5aa478503a7ff3f9e654bd2078ebab)
```
CMake:
```
$ cmake --version
cmake version 3.12.3
```
Compiler: gcc
```
$ g++ --version
g++ (Debian 8.2.0-8) 8.2.0
```

## Steps to reproduce

Clone SFML repository, checkout version 2.5.1, put the following code in `examples/pong/Pong.cpp`:
```cpp
#include <SFML/Graphics.hpp>
#include <iostream>

int main()
{
    sf::RenderWindow window(sf::VideoMode(800, 600), "TEST");
    window.setFramerateLimit(60);

    while(window.isOpen())
    {
        sf::Event event{};
        while(window.pollEvent(event)) {
            if((event.type == sf::Event::KeyPressed) || (event.type == sf::Event::KeyReleased)){
                std::cout << ((event.type == sf::Event::KeyPressed) ? "KeyPressed " : "KeyReleased ");
                std::cout << event.key.code;
                std::cout << " alt=" << event.key.alt;
                std::cout << " control=" << event.key.control;
                std::cout << " shift=" << event.key.shift;
                std::cout << " system=" << event.key.system;
                std::cout << std::endl;
            }

            if(event.type == sf::Event::Closed)
                window.close();
        }

        window.clear();
        window.display();
    }
    return EXIT_SUCCESS;
}
```
Compile pong example with CMake, launch the example, a black window open, press left shift (and release), close the window.

Note: Left shift is sf::Keyboard::Key::LShift = 38

## Expected behavior/output
```
/GitHub/SFML/cmake-build-debug/examples/pong/pong-d
KeyPressed 38 alt=0 control=0 shift=1 system=0
KeyReleased 38 alt=0 control=0 shift=0 system=0
```

## Actual behavior/output
```
/GitHub/SFML/cmake-build-debug/examples/pong/pong-d
KeyPressed 38 alt=0 control=0 shift=0 system=0
KeyReleased 38 alt=0 control=0 shift=1 system=0
```

**Comment by kiwon0905** [comment]
I think the current behavior is actually correct. The shift/alt/ctrl flags reflect the keyboard state when the event is generated. When a shift is pressed, the shift had not been pressed before, so the shift flag isn't set. When a shift is released however, the shift had been pressed, so the shift flag is set.

**Comment by fahlmant** [comment]
If we're going by the documentation, the data type is a bool, and the description is `Is the shift/alt/ctrl key pressed?` Semantically, I read this as asking "Is the key being pressed down at the time of polling?" If that is the question that this even it answering, then the actual behaviour according to this issue is wrong. Otherwise, the documentation should be clarifies to mean "Has the key been pressed and released?" I'm not sure which one is the intended behavior, do any of the SFML team know?

**Comment by LaurentGomila** [comment]
The thing is, it is really up to the operating system. So you may even have inconsistent results across different OSes.

Anyway, a more precise definition of these booleans would be "was the shift/alt/ctrl modifier active (pressed) at the time the event was triggered?".

**Comment by eXpl0it3r** [comment]
May potentially be better handled with scancode support #1235

**Comment by JonnyPtn** [comment]
I don't think scancodes will affect this - seems a subjective matter as to whether the modifier keys are "pressed" at the point they're being pressed. Personally I don't think there's any issue as we're just going by what the OS tells us, at most this is about clarifying the documentation

**Comment by eXpl0it3r** [comment]
Yes, having looked at all the existing well-known implementation, it also seems like everyone does it a bit different and of course it also depends on the OS.
It's something I want to look into after scancodes to define/document how things are on the OS level and how we do it in SFML

**Comment by dgcor** [comment]
I have a similar problem using the latest 2.6 branch with the duplicate key fix applied [here](https://github.com/dgcor/SFML/tree/feature/scancode).

In windows, I have to apply this fix:

https://github.com/dgcor/DGEngine.core/blob/0ec8f4575c6e279732e5be6561e15e607acfe877/src/Parser/ParseInputEvent.cpp#L12

because I'm using the event as a key in a map on the OnKeyPressed event, but that event's struct has the alt key set to true under windows and set to false under linux.

This fix I apply under windows makes it so that the key in the map has the key/scancode set but also the bool value set, since the windows implementation sets both and the linux only sets the key/scancode.
